src/main/java/org/folio/helper/PurchaseOrderLineHelper.java

```diff
@@ -896,6 +899,51 @@ private CompletableFuture<Void> processPoLineEncumbrances(CompositePurchaseOrder
     return completedFuture(null);
   }
 
+  private CompletableFuture<Void> updateInventoryItemStatus(CompositePoLine compOrderLine, JsonObject lineFromStorage, RequestContext requestContext) {
+    PoLine poLineFromStorage =  lineFromStorage.mapTo(PoLine.class);
+    if (isStatusChanged(compOrderLine, poLineFromStorage) && isCurrentStatusCanceled(compOrderLine)) {
+      return purchaseOrderLineService.getPoLinesByOrderId(compOrderLine.getPurchaseOrderId(), requestContext)
+        .thenCompose(poLines -> {
+          if (poLines.size() > 1) {
+            List<String> poLineIds = poLines.stream().map(PoLine::getId).collect(toList());
+            return inventoryManager.getItemsByPoLineIdsAndStatus(poLineIds, ItemStatus.ON_ORDER.value(), requestContext)
+              .thenCompose(items -> {
+                if (items.size() > 1) {
+                  //Each poLine can have only one linked item
+                  Optional<JsonObject> poLineItem = items.stream()
+                    .filter(item -> compOrderLine.getId().equals(item.getString("purchaseOrderLineIdentifier"))).findFirst();
+                  if (poLineItem.isPresent()) {
+                    JsonObject updatedItem = updateItemStatus(poLineItem.get(), ItemStatus.ORDER_CLOSED);
+                    inventoryManager.updateItem(updatedItem, requestContext);
+                  }
+                }
+                return CompletableFuture.completedFuture(null);
+              });
+          }
+          return CompletableFuture.completedFuture(null);
+        });
+    }
+    return CompletableFuture.completedFuture(null);
+  }
+
+  private boolean isStatusChanged(CompositePoLine compOrderLine, PoLine lineFromStorage) {
+    return !StringUtils.equals(lineFromStorage.getReceiptStatus().value(), compOrderLine.getReceiptStatus().value()) ||
+      !StringUtils.equals(lineFromStorage.getPaymentStatus().value(), compOrderLine.getPaymentStatus().value());
+  }
+
+  private boolean isCurrentStatusCanceled(CompositePoLine compOrderLine) {
+    return CompositePoLine.ReceiptStatus.CANCELLED.equals(compOrderLine.getReceiptStatus()) ||
+      CompositePoLine.PaymentStatus.CANCELLED.equals(compOrderLine.getPaymentStatus());
+  }
+
+  private JsonObject updateItemStatus(JsonObject poLineItem, ItemStatus itemStatus) {
+    JsonObject status = new JsonObject();
+    status.put("name", itemStatus);
```

Comment:
nitpick: feeling like this logic should be somewhere with some other Inventory information. Maybe `inventoryManager`.